### DISCUSSION TITLE
Upgrade version of OpenTofu in versions.tf

### DISCUSSION THREAD

**Discussion by muffl0n** [renovatebot/renovate#29786] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gitlab. latest

### Please tell us more about your question or problem

We're just starting to switch from Terraform to OpenTofu and would like Renovate to upgrade the `required_version` which is defined in `versions.tf`:
```
terraform {
  required_version = "1.7.0"
  required_providers {
    google      = "5.32.0"
  }
}
```

To tell Renovate to not upgrade to the latest version of Terraform but to the latest version of OpenTofu we changed the `renovate.json` accordingly:
```
{
  ...
  "packageRules": [
    {
      "matchDatasources": ["github-releases"],
      "matchDepNames": ["hashicorp/terraform"],
      "enabled": false
    }
  ],
  "customManagers": [
    {
      "customType": "regex",
      "fileMatch": ["^versions.tf$"],
      "matchStrings": ["required_version\\s=\\s\"(?<currentValue>.*?)\""],
      "depNameTemplate": "opentofu/opentofu",
      "datasourceTemplate": "github-releases"
    }
  ]
}
```
So we disable the upgrade for package `hashicorp/terraform` and add a new `customManager` to upgrade `opentofu/opentofu`.

Though it seem to work, it's a bit cumbersome. Is there a better way? 

I found https://docs.renovatebot.com/modules/manager/terraform/#terraform-vs-opentofu but this seems to only work for modules, right?

### Logs (if relevant)

<details><summary>Logs</summary>

```
{
  "name": "renovate",
  "hostname": "runner-kgoikxeqv-project-44506035-concurrent-5-s5q0yhwd",
  "pid": 36,
  "level": 20,
  "logContext": "QTpQ0-PMogWwiCVFnEh15",
  "repository": "terraform-renovatebot",
  "baseBranch": "main",
  "config": {
    "terraform": [
      {
        "deps": [
          {
            "currentValue": "5.32.0",
            "depType": "required_provider",
            "depName": "google",
            "datasource": "terraform-provider",
            "packageName": "hashicorp/google",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "5.34.0",
                "newValue": "5.34.0",
                "releaseTimestamp": "2024-06-17T17:31:10.000Z",
                "newMajor": 5,
                "newMinor": 34,
                "newPatch": 0,
                "updateType": "minor",
                "branchName": "renovate/755b31b43be2a871d058309ceb"
              }
            ],
            "versioning": "hashicorp",
            "warnings": [],
            "sourceUrl": "https://github.com/hashicorp/terraform-provider-google",
            "registryUrl": "https://registry.terraform.io",
            "homepage": "https://registry.terraform.io/providers/hashicorp/google",
            "currentVersion": "5.32.0",
            "isSingleVersion": true,
            "fixedVersion": "5.32.0"
          },
          {
            "currentValue": "1.7.0",
            "depType": "required_version",
            "datasource": "github-releases",
            "depName": "hashicorp/terraform",
            "extractVersion": "v(?<version>.*)$",
            "versioning": "hashicorp",
            "updates": [],
            "packageName": "hashicorp/terraform",
            "skipReason": "disabled"
          }
        ],
        "packageFile": "versions.tf"
      }
    ],
    "regex": [
      {
        "deps": [
          {
            "depName": "opentofu/opentofu",
            "currentValue": "1.7.0",
            "datasource": "github-releases",
            "replaceString": "required_version = \"1.7.0\"",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "v1.7.2",
                "newValue": "1.7.2",
                "releaseTimestamp": "2024-06-03T15:36:53.000Z",
                "newMajor": 1,
                "newMinor": 7,
                "newPatch": 2,
                "updateType": "patch",
                "branchName": "renovate/76f9c04a90c4e52f241c6ddb5a"
              }
            ],
            "packageName": "opentofu/opentofu",
            "versioning": "semver-coerced",
            "warnings": [],
            "sourceUrl": "https://github.com/opentofu/opentofu",
            "registryUrl": "https://github.com",
            "currentVersion": "v1.7.0",
            "currentVersionTimestamp": "2024-04-30T15:44:22.000Z",
            "isSingleVersion": true,
            "fixedVersion": "1.7.0"
          }
        ],
        "matchStrings": [
          "required_version\\s=\\s\"(?<currentValue>.*?)\""
        ],
        "depNameTemplate": "opentofu/opentofu",
        "datasourceTemplate": "github-releases",
        "packageFile": "versions.tf"
      }
    ]
  },
  "msg": "packageFiles with updates",
  "time": "2024-06-20T20:05:33.608Z",
  "v": 0
}
```

</details>

**Comment by secustor** [comment, marked as answer]
The guide is for providers only. Modules shouldn't be needing any changes. 

For now there is no easier solution. 
Though if the popularity keeps increasing we could wrap the `terraform` manager and provide this functionality.

**Reply by muffl0n** [reply]
Thank you very much! I ended up with this preset-file:
```
{
  "packageRules": [
    {
      "matchDatasources": ["github-releases"],
      "matchDepNames": ["hashicorp/terraform"],
      "enabled": false
    }
  ],
  "customManagers": [
    {
      "customType": "regex",
      "fileMatch": ["^versions.tf$"],
      "matchStrings": ["required_version\\s=\\s\"(?<currentValue>.*?)\""],
      "depNameTemplate": "opentofu/opentofu",
      "datasourceTemplate": "github-releases"
    }
  ]
}
```

**Reply by rarkins** [reply]
Is there any way for renovate to determine from the repo files whether opentofu is in use?

**Reply by muffl0n** [reply]
The way we distinguish this in one of our scripts is to check for the presence of `.terraform-version` vs `.opentofu-version` like they are used for `tfenv` and `tofuenv`.
But I'm not sure if that is common enough to rely on.

**Reply by muffl0n** [reply]
There's https://github.com/opentofu/opentofu/issues/1328 which allow files to end in `.otf`. We should probably wait for this release. Then we could just add an OpenTofu manager that handles upgrade of `required_version` in `versions.otf`.

**Reply by HonkingGoose** [reply]
> There's [opentofu/opentofu#1328](https://github.com/opentofu/opentofu/issues/1328) which allow files to end in `.otf`. We should probably wait for this release. Then we could just add an OpenTofu manager that handles upgrade of `required_version` in `versions.otf`.

That linked issue has been closed as resolved. I think this upstream PR is related to the closing of that issue:

- https://github.com/opentofu/opentofu/pull/1738

I don't see anything about `.otf` files in the linked PR, but it looks like `.tofu` file endings are now supported by OpenTofu? So maybe we can tell Renovate users to switch to `.tofu`?

**Reply by ghost** [reply]
Hey folks, OpenTofu core dev here, the `.otf` file extension has changed to `.tofu` during the development process. Starting with OpenTofu 1.8, the presence of a file ending in `.tofu` or any of its alternatives would be a good indicator that OpenTofu is in use, but it doesn't mean that the project in question isn't compatible with both OpenTofu and Terraform.

The ideal implementation would take the `terraform` &rarr; `required_version` present in a `.tofu` file as the required OpenTofu version and a `terraform` &rarr; `required_version` in a `.tf` file as a required Terraform version with the option for the user to manually force one or the other.

**Reply by HonkingGoose** [reply]
## Add OpenTofu-only mode to Renovate?

Maybe we could add a special toggle to Renovate, to put it in "OpenTofu only-mode"? Here's how that would work:

1. User puts `openTofuOnlyMode: true` or `openTofuOnlyMode=enabled` in their Renovate config file.
2. When Renovate runs, it _only_ uses the _OpenTofu_ program to interpret the relevant files.
3. When Renovate runs, it will read `required_version` from `.tf` files as meaning _use x version of OpenTofu_. Or in simpler words: all `required_version` in `.tofu` and `.tf` files are OpenTofu versions.
4. Renovate will prefer to follow rules/conventions/guidelines that are common to OpenTofu.

## Add a proper `opentofu` manager to Renovate?

Or maybe we should add a proper `opentofu` manager to Renovate? That way Renovate users can enable the `opentofu` manager, and disable the `terraform` manager in their config.

For Renovate to automatically pick the right manager, we probably need this feature first:

- #29809

**Reply by secustor** [reply]
The way to go is a separate `opentofu` manager which reuses logic  `terraform` and is enabled for `.tofu` files and turned off for `.tf` by default. 

With the addition of #29809 we can dynamically turn off `terraform` and add `.tf` files to `opentofu` if we detect it.

**Reply by HonkingGoose** [reply]
I'll let one of you maintainers create a issue for the new `opentofu` manager then. :wink: 

**Edit:** Nevermind, I created an issue for us. I copied `@secustor`'s comment and cleaned it up a bit. See:

- #30500
